File: rrag-pipeline-service/src/pipeline_service/models/pipeline.py
from __future__ import annotations

import enum
import uuid

from sqlalchemy import JSON, Enum as SAEnum, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base, TimestampMixin


class PipelineStatus(str, enum.Enum):
    draft = "draft"
    published = "published"
    archived = "archived"


class PipelineType(str, enum.Enum):
    ingestion = "ingestion"
    query = "query"


class Pipeline(TimestampMixin, Base):
    __tablename__ = "pipelines"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    workspace_id: Mapped[str] = mapped_column(String(36), index=True)
    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    definition: Mapped[dict] = mapped_column(JSON, default=dict)
    status: Mapped[PipelineStatus] = mapped_column(
        SAEnum(PipelineStatus), default=PipelineStatus.draft
    )
    pipeline_type: Mapped[PipelineType] = mapped_column(
        SAEnum(PipelineType), default=PipelineType.ingestion, index=True
    )
    datastore_id: Mapped[str | None] = mapped_column(
        String(36), nullable=True, index=True
    )
    created_by: Mapped[str] = mapped_column(String(36))

    versions: Mapped[list[PipelineVersion]] = relationship(
        back_populates="pipeline", cascade="all, delete-orphan"
    )
    runs: Mapped[list["PipelineRun"]] = relationship(
        back_populates="pipeline", cascade="all, delete-orphan"
    )


class PipelineVersion(TimestampMixin, Base):
    __tablename__ = "pipeline_versions"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    pipeline_id: Mapped[str] = mapped_column(
        ForeignKey("pipelines.id", ondelete="CASCADE")
    )
    version: Mapped[int] = mapped_column(Integer)
    definition: Mapped[dict] = mapped_column(JSON)
    change_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_by: Mapped[str] = mapped_column(String(36))

    pipeline: Mapped[Pipeline] = relationship(back_populates="versions")

File: rrag-frontend/src/types/pipeline.ts
import type { Node, Edge } from '@xyflow/react';

export interface PipelineNodeData {
  label: string;
  componentType: string;
  category: string;
  config: Record<string, unknown>;
  [key: string]: unknown;
}

export type PipelineNode = Node<PipelineNodeData>;
export type PipelineEdge = Edge;

export type PipelineType = 'ingestion' | 'query';

export interface Pipeline {
  id: string;
  workspaceId: string;
  name: string;
  description: string | null;
  definition: { nodes: PipelineNode[]; edges: PipelineEdge[] };
  status: 'draft' | 'published' | 'archived';
  pipelineType: PipelineType;
  datastoreId: string | null;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface PipelineVersion {
  id: string;
  pipelineId: string;
  version: number;
  definition: Record<string, unknown>;
  changeMessage: string | null;
  createdBy: string;
  createdAt: string;
}

export interface PipelineRun {
  id: string;
  pipelineId: string;
  versionId: string | null;
  status: 'pending' | 'running' | 'completed' | 'failed';
  inputParams: Record<string, unknown> | null;
  result: Record<string, unknown> | null;
  error: string | null;
  startedAt: string | null;
  completedAt: string | null;
  createdBy: string;
  createdAt: string;
}

export interface ConfigSchemaProperty {
  type: string;
  default?: unknown;
  description?: string;
  enum?: string[];
  minimum?: number;
  maximum?: number;
  items?: { type?: string; enum?: string[] };
  ui_widget?: 'text' | 'textarea' | 'password' | 'select' | 'slider' | 'toggle' | 'array' | 'json';
  ui_advanced?: boolean;
  ui_group?: string;
}

export interface ConfigSchema {
  type: string;
  properties?: Record<string, ConfigSchemaProperty>;
  required?: string[];
}

export interface StepDefinition {
  order: number;
  name: string;
  method: string;
  description?: string;
  retry?: number;
  retryCondition?: string;
  outputs?: string[];
  configSchema?: ConfigSchema;
}

export interface ComponentDefinition {
  type: string;
  category: string;
  name: string;
  description: string;
  configSchema: ConfigSchema;
  inputPorts: { name: string; type: string; description?: string }[];
  outputPorts: { name: string; type: string; description?: string }[];
  source?: 'seed' | 'custom';
  isComposite?: boolean;
  steps?: StepDefinition[];
  workspaceId?: string | null;
}

export interface CompositeNodeData extends PipelineNodeData {
  isComposite: true;
  steps: StepDefinition[];
  isExpanded: boolean;
}
